cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/Dag.java

```diff
@@ -46,9 +46,11 @@
   protected final SetMultimap<String, String> outgoingConnections;
   // stage -> inputs for that stage
   protected final SetMultimap<String, String> incomingConnections;
+  protected final Set<Connection> connections;
 
   public Dag(Collection<Connection> connections) {
     Preconditions.checkArgument(!connections.isEmpty(), "Cannot create a DAG without any connections");
+    this.connections = new HashSet<>(connections);
```

Comment:
seems like this is only being added in order to eventually get the connections again in order to build a ConnectorDag.  Instead of introducing this redundant set, I think it's better to add a constructor to ConnectorDag that takes a Dag instead of a `Collection<Connection>`, and uses the Dag constructor with its outgoing and incoming connections. The builder can also have an addDag() method that can be used instead of addConnection()